src/java.base/share/classes/java/security/Provider.java

```diff
@@ -749,196 +1464,195 @@ private void readObject(ObjectInputStream in)
             // otherwise, set version based on versionStr
             this.version = parseVersionStr(this.versionStr);
         }
-        this.serviceMap = new ConcurrentHashMap<>();
-        this.legacyMap = new ConcurrentHashMap<>();
+        this.servicesMap = new ServicesMap();
         this.prngAlgos = new LinkedHashSet<>(6);
         implClear();
         initialized = true;
         putAll(copy);
     }
 
-    // returns false if no update necessary, i.e. key isn't String or
-    // is String, but it's provider-related (name/version/info/className)
-    private static boolean checkLegacy(Object key) {
-        if (key instanceof String && ((String)key).startsWith("Provider.")) {
-            // ignore provider related updates
-            return false;
-        } else {
-            return true;
+    /*
+     * Enum to determine if changes to the Properties map must be applied by the
+     * caller (UPDATE) or skipped (SKIP).
+     *
+     * If a change does not concern a ServicesMap, UPDATE is returned. An
+     * example of this is when adding, modifying or removing an entry that is
+     * not a service, alias or attribute.
+     *
+     * If the change concerns a ServicesMap, SKIP is returned. The change may
+     * have been applied internally or ignored due to an error. In the former
+     * case, Properties map entries are synchronized. In the latter, Properties
+     * map entries are not modified.
+     */
+    private enum PropertiesMapAction {
+        UPDATE,
+        SKIP
+    }
+
+    private PropertiesMapAction doLegacyOp(ServicesMap servicesMap, Object key,
+            Object value, Object oldValue, OPType opType) {
+        if (key instanceof String ks) {
+            if (ks.startsWith("Provider.")) {
+                // Ignore provider related updates.
+                return PropertiesMapAction.SKIP;
+            }
+            if (value instanceof String vs) {
+                return parseLegacy(servicesMap, ks, vs, opType);
+            } else if (value != null && oldValue instanceof String oldValueS &&
+                    opType == OPType.ADD) {
```

Comment:
Which method is this else-block for? `value` is not null and not instanceof `String` and `oldValue` is instanceof `String` and `opType` is ADD?

Reply:
In this case we are adding an entry to the Provider (seen as a Properties map more than a Provider). The value of this new entry is not a string and is replacing an entry with the same key, whose value is mapped to a service in the internal (services) map. From the point of view of the internal map, this acts as a removal.